grove/tomography/tomography.py

```diff
@@ -0,0 +1,727 @@
+        ]
+
+        rho_m_real_imag = sum((rm * ut.to_realimag(Pm)
+                               for (rm, Pm) in izip(rho_m, pauli_basis.ops)), 0)
+
+        if POSITIVE in settings.constraints:
+            if _SDP_SOLVER.is_functional():
+                constraints.append(rho_m_real_imag >> 0)
+            else:  # pragma no coverage
+                _log.warning("No convex solver capable of semi-definite problems installed.\n"
+                             "Dropping the positivity constraint on the density matrix")
+
+        if UNIT_TRACE in settings.constraints:
+            # this assumes that the first element of the Pauli basis is always proportional to
+            # the identity
+            constraints.append(rho_m[0, 0] == 1. / pauli_basis.ops[0].tr().real)
+
+        prob = cvxpy.Problem(cvxpy.Minimize(obj), constraints)
+
+        _log.info("Starting convex solver")
+        prob.solve(solver=SOLVER, **settings.solver_kwargs)
+        if prob.status != cvxpy.OPTIMAL:  # pragma no coverage
+            _log.warning("Problem did not converge to optimal solution. "
+                         "Solver settings: {}".format(settings.solver_kwargs))
+
+        return StateTomography(np.array(rho_m.value).ravel(), pauli_basis, settings)
+
+    def __init__(self, rho_coeffs, pauli_basis, settings):
+        """
+        Construct a StateTomography to encapsulate the result of estimating the quantum state from
+        a quantum tomography measurement.
+
+        :param numpy.ndarray r_est: The estimated quantum state represented in a given (generalized)
+         Pauli basis.
+        :param OperatorBasis pauli_basis: The employed (generalized) Pauli basis.
+        :param TomographySettings settings: The settings used to estimate the state.
+        """
+        self.rho_coeffs = rho_coeffs
+        self.pauli_basis = pauli_basis
+        self.rho_est = sum((r_m * p_m for r_m, p_m in izip(rho_coeffs, pauli_basis.ops)))
+        self.settings = settings
+
+    def fidelity(self, other):
+        """
+        Compute the quantum state fidelity of the estimated state with another state.
+
+        :param qutip.Qobj other: The other quantum state.
+        :return: The fidelity, a real number between 0 and 1.
+        :rtype: float
+        """
+        return qt.fidelity(self.rho_est, other)
+
+    def plot_state_histogram(self, ax):
+        """
+        Visualize the complex matrix elements of the estimated state.
+
+        :param matplotlib.Axes ax: A matplotlib Axes object to plot into.
+        """
+        title = "Estimated state"
+        nqc = int(round(np.log2(self.rho_est.data.shape[0])))
+        labels = ut.basis_labels(nqc)
+        return ut.state_histogram(self.rho_est, ax, title)
+
+    def plot(self):
+        """
+        Visualize the state.
+
+        :return: The generated figure.
+        :rtype: matplotlib.Figure
+        """
+        f = plt.figure(figsize=(10, 8))
```

Comment:
Nitpick for design ... make it:
```
width = 10
height = width / 1.618
```
to implement the pleasing golden ratio :)

Reply:
lol :+1: